chia/wallet/puzzles/singleton_top_layer.py

```diff
@@ -20,6 +20,138 @@
+#         |
+#  ------------------ Atomic Transaction 1 -----------------
+#         v
+#   .------------.       .-------------------------------.
+#   | Launcher   |------>| Eve Coin Containing Program I |
+#   `------------'       `-------------------------------'
+#                                        |
+#  -------------------- End Transaction 1 ------------------\
+#                                        |                   > The Eve coin
+#  --------------- (2) Transaction With I ------------------/  may also be
+#                                        |                     spent
+#                                        v                     simultaneously
+#                 .-----------------------------------.
+#                 | Running Singleton With Program I' |
+#                 `-----------------------------------'
+#                                        |
+# --------------------- End Transaction 2 ------------------
+#                                        |
+# --------------- (3) Transaction With I' ------------------
+# ...
+#
+#
+# == Practial use of singleton_top_layer.py ==
+#
+# 1) Designate some coin as coin A
+#
+# 2) call puzzle_for_singleton with that coin's name (it is the Parent of the
+#    Launch coin), and the initial inner puzzle I, curried as appropriate for
+#    its own purpose. Adaptations of the program I and its descendants are
+#    required as below.
+#
+# 3) call launch_conditions_and_coinsol to get a set of "launch_conditions",
+#    which will be used to spend standard coin A, and a "spend", which spends
+#    the Launcher created by the application of "launch_conditions" to A in a
+#    spend. These actions must be done in the same spend bundle.
+#
+#    One can create a SpendBundle containing the spend of A giving it the
+#    argument list (() (q . launch_conditions) ()) and then append "spend" onto
+#    its .coin_spends to create a combined spend bundle.
+#
+# 4) submit the combine spend bundle.
+#
+# 5) Remember the identity of the Launcher coin:
+#
+#      Coin(A.name(), SINGLETON_LAUNCHER_HASH, amount)
+#
+# A singleton has been created like this:
+#
+#      Coin(Launcher.name(), puzzle_for_singleton(Launcher.name(), I), amount)
+#
+#
+# == To spend the singleton requires some host side setup ==
+#
+# The singleton adds an ASSERT_MY_COIN_ID to constrain it to the coin that
+# matches its own conception of itself.  It consumes a "LineageProof" object
+# when spent that must be constructed so.  We'll call the singleton we intend
+# to spend "S".
+#
+# Specifically, the required puzzle is the Inner puzzle I for the parent of S
+# unless S is the Eve coin, in which case it is None.
+# So to spend S', the second singleton, I is used, and to spend S'', I' is used.
+# We'll call this puzzle hash (or None) PH.
+#
+#      If this is the Eve singleton:
+#
+#          PH = None
+#          L = LineageProof(Launcher, PH, amount)
+#
+#      Otherwise
+#
+#          PH = ParentOf(S).puzzle_hash
```

Comment:
I think this should be `PH = ParentOf(S).inner_puzzle_hash`

Reply:
ya.  making this change.